omniauth-ebay-oauth implementation

ihnatz opened the pull request.

ihnatz committed:
Add OmniAuth eBay strategy gem basic structure

ihnatz committed:
Configure first two OAuth steps

ihnatz committed:
Add user info OmniAuth step

ihnatz committed:
Add option to run OmniAuth in sandbox mode

ihnatz committed:
UerInfoRequest -> UserInfo to separate concerns

ihnatz committed:
Add valid eBay user to omniauth hash mapping

ihnatz committed:
Change name to satisfy Strategy Contribution Guide

ihnatz committed:
Add rubocop to default rake task

ihnatz committed:
Add simplecov coverage

ihnatz committed:
Add read timout for http requests

ihnatz committed:
Update README with information

ihnatz committed:
Add last version to travis config

ihnatz committed:
Minor refactoring tip

palkan commented on spec/omniauth/ebay-oauth/user_info_request_spec.rb:
```diff
@@ -0,0 +1,43 @@
+  let(:failure_result)  { File.read('spec/fixtures/result_failure.xml') }
+  let(:success_result)  { File.read('spec/fixtures/result_success.xml') }
+
+  before { stub_const("#{described_class}::USER_REQUEST", body) }
```
Why do we have to stub this const and not use it as `body`?

ihnatz replied:
The main reason -- it's easier to read when something went wrong, compare:

```
       You can stub this request with the following snippet:

       stub_request(:post, "https://api.com/endpoint").
         with(:body => "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<GetUserRequest xmlns=\"urn:ebay:apis:eBLBaseComponents\">\n  <ErrorLanguage>en_US</ErrorLanguage>\n  <WarningLevel>High</WarningLevel>\n  <DetailLevel>ReturnAll</DetailLevel>\n</GetUserRequest>\n",
              :headers => {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Content-Type'=>'text/xml', 'Host'=>'api.com', 'User-Agent'=>'Ruby', 'X-Ebay-Api-Call-Name'=>'GetUser', 'X-Ebay-Api-Compatibility-Level'=>'967', 'X-Ebay-Api-Iaf-Token'=>'token+1', 'X-Ebay-Api-Siteid'=>'0'}).
         to_return(:status => 200, :body => "", :headers => {})

       registered request stubs:

       stub_request(:post, "https://api.com/endpoint").
         with(:body => "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<GetUserRequest xmlns=\"urn:ebay:apis:eBLBaseComponents\">\n  <ErrorLanguage>en_US</ErrorLanguage>\n  <WarningLevel>High</WarningLevel>\n  <DetailLevel>ReturnAll</DetailLevel>\n</GetUserRequest>\n",
              :headers => {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Content-Type'=>'text/xml', 'Host'=>'api.com', 'User-Agent'=>'Ruby', 'X-Ebay-Api-Call-Name'=>'GetUser', 'X-Ebay-Api-Compatibility-Level'=>'967', 'X-Ebay-Api-Iaf-Token'=>'token', 'X-Ebay-Api-Siteid'=>'0'})
```

vs

```
       You can stub this request with the following snippet:

       stub_request(:post, "https://api.com/endpoint").
         with(:body => "<>",
              :headers => {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Content-Type'=>'text/xml', 'Host'=>'api.com', 'User-Agent'=>'Ruby', 'X-Ebay-Api-Call-Name'=>'GetUser', 'X-Ebay-Api-Compatibility-Level'=>'967', 'X-Ebay-Api-Iaf-Token'=>'token+1', 'X-Ebay-Api-Siteid'=>'0'}).
         to_return(:status => 200, :body => "", :headers => {})

       registered request stubs:

       stub_request(:post, "https://api.com/endpoint").
         with(:body => "<>",
              :headers => {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Content-Type'=>'text/xml', 'Host'=>'api.com', 'User-Agent'=>'Ruby', 'X-Ebay-Api-Call-Name'=>'GetUser', 'X-Ebay-Api-Compatibility-Level'=>'967', 'X-Ebay-Api-Iaf-Token'=>'token', 'X-Ebay-Api-Siteid'=>'0'})
```

Seems like it will be nice even stub default headers, to reduce all this mass. Furthermore, we don't depend on this constants as code control structure, so we could stub it for free.

BTW, changed from `stub_const` to DI.

ihnatz committed:
Use DI instead of const stubing

Envek commented on lib/omniauth/ebay-oauth/user_info.rb:
```diff
@@ -0,0 +1,47 @@
+    # https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema
+    class UserInfo
+      MAPPING = {
+        uid: %w[GetUserResponse User UserID],
```
Unfortunately, eBay username is not reliable value as the user can change it and can change it frequently.

So it's better to use scary `EIASToken` value because it doesn't change. Username is still important though, so it's better to keep it in `nickname` field of info hash as per https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema#schema-10-and-later

ihnatz replied:
Fixed, UserID moved to nickname

Envek commented on lib/omniauth/ebay-oauth/user_info_request.rb:
```diff
@@ -0,0 +1,74 @@
+      private
+
+      def ensure_success_code
+        lambda { |response|
```
Style note: you can get rid of lambdas and use just methods by using `method` method:

```rb
def call
  MultiXml.parse(…).tap(&method(:ensure_success_result))
end

def ensure_success_code(response)
 …
end
```

ihnatz replied:
A matter of taste, but changed to this way

Envek commented on lib/omniauth/ebay-oauth/user_info_request.rb:
```diff
@@ -0,0 +1,74 @@
+        Net::HTTP.new(@url.host, @url.port).tap do |http|
+          http.read_timeout = @read_timeout
+          http.use_ssl = true
+          http.verify_mode = OpenSSL::SSL::VERIFY_NONE
```
Please don't do that! This is a security hole!

You may wish to enable bypassing SSL certificate check via an option, but that option must be disabled by default.

ihnatz replied:
Yep, agree, shame on me, fixed

Envek commented on README.md:
```diff
@@ -9,17 +9,63 @@ OmniAuth Strategy for eBay Apps (for using with eBay REST APIs)
+
+Additional options:
+  - __sandbox__ - Are you running your application in [sandbox mode](<https://developer.ebay.com/api-docs/static/sandbox-landing.html>), default __`true`__.
+  - __scope__ - A list of [OAuth scopes](<https://developer.ebay.com/api-docs/static/oauth-details.html#scopes>) that provide access to the interfaces you call.
```
Maybe it worth to note that by default that list of scopes is empty and users should provide it if they will use eBay APIs. Or perhaps it will be good to include public access scope `https://api.ebay.com/oauth/api_scope` by default.

ihnatz replied:
Not sure about default value, it's omniauth strategy and eBay returns enough information for simple SSO. Add ability to pass scope as array of scopes or string.

Envek reviewed (changes requested):
Thank you for quick and clean implementation!

Some things that should be fixed for use in real apps:

 1. eBay username. It's unreliable to use it as user identifier as it may be changed by the user.

    See the code comment for details.

 2. In real apps, it is important to know when `refresh_token` will expire to send a reminder to the user, asking to log in again. Also, it allows distinguishing between token expiration and token revocation by the user.

    Please add `refresh_token_expires_at` field into `credentials` hash.

 3. OAuth scopes. Empty scope list is fine for sign-in purposes, but it should be documented better IMHO. 

    Also, an example should be added (should user concatenate the list of scopes with space or can we do it for them?)

ihnatz committed:
Change uid to nickname, new field for uid

ihnatz committed:
Use methods reference instead of lambda methods

ihnatz committed:
Remove VERIFY_NONE ssl mode

ihnatz committed:
Add ability to pass scopes as array

ihnatz committed:
Add refresh_token_expires_at to credentials hash

ihnatz committed:
Change const to ref to allow inheritance chain

ihnatz committed:
Fix fragile test with timing

Spec
    Start     Time.now       Expiration time
      |          |___expiration___|
      v          v----------------v
|-----------------------------------| t
       ^       ^----------------^
       |       |___expiration___|
     Start  Time.now       Expiration time
Code

Envek commented:
Released as v0.1.0. Congratulations and thank you for your work!
